Lua Select API, Lua based KeySelect & Customisable Smart Select

whossname opened the pull request:
### Work done

Created a Lua based replacement for the existing hotkey selection code and added functionality to SmartSelect to utilise the same API to modify which units are selected when using drag select.

There were two main motivations for this:

- the existing C++ API can be inconsistent and buggy in places. As far as I can tell the NameContains and Category rules are broken, and the select count/one/part conclusions includes units that are already selected, which can be counter intuitive.

- writing it in Lua allows the code to be updated, maintained and improved more easily. Far more people should be able to work on the Lua code, and it should be easier to apply the changes.

An additional driver was being able to use the select syntax for SmartSelect commands in uikeys.txt.

The following filter was not implemented, I have no idea what it is meant to do:  `RulesParamEquals_<string>_<integer>`

I was advised that no one uses it anyway. Also an extra rule `AntiAir` was added for selecting Anti-air units.

The relevant files are:
- `select_api.lua` the common code used in both SmartSelect and KeySelect to decode selection command strings and cache the resulting select function
- `unit_smart_select.lua` this widget was modified to use the select api for selectbox commands in uikeys.txt
- `unit_key_select.lua` intended as a replacement for the existing select commands. This widget is disabled by default as we want users to test it before it is fully rolled out as a replacement.
- `compare_to_spring.lua` a test script written to identify differences between the existing C++ select implementation and the new Lua select implementation. It's not a proper unit/integration test as the results require a lot of interpretation.

See discord thread for more details: https://discord.com/channels/549281623154229250/1266331814440730677

#### Setup
The changes to SmartSelect are merged in directly. It will require modification to the uikeys.txt file to use. Something like this:

```
bind          Any+shift  selectbox Not_Aircraft
bind           Any+ctrl  selectbox Buildoptions
bind            Any+alt  selectbox Aircraft
bind          Any+space  selectbox Builder_Aircraft
```

To test the KeySelect open widgets and enable the KeySelect widget.

#### Test steps
SmartSelect
- Copy the above code into uikeys.txt and run `/keyreload`
- Create a variety of air/land units
- The `shift`, `ctrl`, `alt` and `space` modifiers should change which units are selected
- Check that the existing SmartSelect drag selection key binds still work as expected

KeySelect
- Run `/wigetselector` and enable `KeySelect`
- Check that all of your keybinds still work

### Known issues

There seems to be an issue where the `NameContain` filter doesn't work as expected if you have `ClearSelection` in the conclusion. It will centre on the unit, but not actually select it. I suspect it also happens for the `Category` filter. No idea why. Seeing as the `NameContain` and `Category` filters don't seem to work correctly in the existing version anyway, I don't see it as a deal breaker.

whossname commented:
The changes to [luamocks/SpringLuaApi.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3757/files#diff-2a8a5eb0c02a0afd9a76f7f51abc1d98f636eacfca37706471f5bd3437eed040) are unintentional. It's mostly removing white-space at the end of lines. I must have accidentally run a formatter at some stage.

sprunk commented on luaui/Widgets/unit_key_select.lua:
```diff
@@ -0,0 +1,29 @@
+		author = "woss (Tyson Buzza)",
+		date = "Aug 24, 2024",
+		license = "Public Domain",
+		layer = -999999,
```
why -999999?

whossname replied:
No idea. I think I was copying the smart select widget. I don't know what that number does.

badosu replied:
You can replace it with 1, the layer is the loading order of the widget in relation to other widgets. We don't expect any conflicts to exist unless a game or user already declares another `select` action in a widget.

badosu commented:
Just for context, the key addition here is the select api and possibility of replacing engine-level select action with an own lua managed one.

Reason is so that select keys can be customized to extend the select api in the future so the game does not depend on engine for it. If, after inclusion and continued use, we verify that all current engine features are satisfied with expected performance profile - this will be eligible for inclusion as engine basecontent and a proposal for engine select removal in the future.

I don't have any particular opinion on the smart select changes, a very small set of players actually even use the current smart select capabilities anyway, I leave that up to the team members to consider.

I'd be very glad if we could include the select api engine and select widget, even if enabled false - to at some point and basic testing enabling to true, and at some point to actually prove this engine is a worthy replacement of the engine select command in basecontent.

Thank you.

badosu commented on luamocks/SpringLuaApi.lua:
```diff
@@ -70,6 +70,7 @@ Engine = {
 
 ---@param path string
+--- @return any
 function VFS.Include(path)
```
Can you revert this file? I understand this file is autogenerated.

whossname replied:
Reverting this line is easy. I don't know where the other changes even came from. It's a few dozen instances of removing the white space at the end of the line.

Out of curiosity, is there a way of setting sensible return types in that file? I think (because this was 2 months ago at this stage) the IDE was yelling at me a lot for mismatching type specs.

whossname replied:
I reverted the white space changes as well.

rhys-vdw replied:
FYI #3919 makes this change unnecessary.

Incidentally, I'd be interested to know how these files are autogenerated. I thought this particular file is hand-written but there is a subfolder of auto-generated files.

WatchTheFort commented:
I need to take a deeper look at the code, but from a first glance it looks like it is just duplicating the engine functionality, i.e. implementing a keybind syntax then using string parsing to turn a text file into Lua select commands.

This seems like the wrong approach, as it is sacrificing the flexibility of Lua. Wouldn't this make more sense as a widget, with the caveat that we need better modding support for custom selection keybinds?

badosu commented:
> Wouldn't this make more sense as a widget...?

Yes, that's what this PR does.

> This seems like the wrong approach, as it is sacrificing the flexibility of Lua. 

That's the opposite, we're using the flexibility of Lua to have the ability to extend engine functionality. It is duplicate, but the plan is to nuke engine select command anyway, this should be a basecontent widget instead suffice it works satisfiably.

If the argument is to then, "if we can do anything then we dont need the old engine select constraints", that's a decent argument. But keep in mind, first supporting base engine select syntax we are able to have this a first step into what will eventually be included in engine basecontent (the testing is priceless for me as an engine developer).

If the game deems this not fruitful for game concerns then sure, that's another arguments, I just ask to give this a chance so we can improve our engine using this testbed.

WatchTheFort commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+	elseif startsWith(sourceDef, "FromMouse_") then
+		return parseNumber(sourceDef, function(distance)
+			local x, y, z = getMouseWorldPos()
+			if z and y and z then
```
```suggestion
			if x and y and z then
```

WatchTheFort commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+				return {}
+			end
+		end)
+	elseif startsWith(sourceDef, "FromMouseC_") then
```
```suggestion
	elseif startsWith(sourceDef, "FromMouseCylinder_") then
```

WatchTheFort commented on luaui/Widgets/Include/select_api.lua:
For numeric filters, do they handle non-integers and negative numbers?

whossname replied:
They all use this pattern:

```
			local minRange = tonumber(getNextToken())
			if not minRange then
				break
			end
```

I'm pretty sure that means it can be a negative float, which doesn't seem like an issue to me. Negative numbers aren't useful, but they don't break the logic. Any non-numbers will just break the parsing.

WatchTheFort replied:
But will negative numbers break game behaviour? Are they ever even meaningful?

whossname replied:
From memory, they don't break behaviour, but they are also never meaningful. Negative numbers behave the same as 0. It either includes all or excludes all.

whossname replied:
Now that I think about it, I'm more worried about how 0 behaves than negative numbers. It will probably be treated as if it's a null.

whossname replied:
I'm explicitly handling nil incase there are 0s now

whossname replied:
tested negative numbers. They behave as expected.

sprunk commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+local SelectApi = {}
+
+-- command definitions from https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Sim/Units/CommandAI/Command.h
+local CMD_STOP = 0
+local CMD_WAIT = 5
+local CMD_PATROL = 15
+local CMD_GUARD = 25
```
yes

whossname replied:
I was looking for that, now that I know what it's called it was easy to find examples.

sprunk commented on luaui/Widgets/Include/select_api.lua:
are they? this is modding support

WatchTheFort replied:
What do you mean? This is pretty standard in other widgets.
```lua
for unitID, unitDef in pairs(UnitDefs) do
   isCommander[unitID] = unitDef.customParams.iscommander
end
```

sprunk replied:
clearly the other widgets suck then

sprunk replied:
ok, i didn't read the code. unit properties aren't static, but unit **def** properties are, which is what this code looks at. so yes, it is a good idea to pre-cache

whossname replied:
Does this mean this one is fine:

```
local function simpleUdefFilter(invert, property)
	return invertCurry(invert, function(udef)
		return udef[property]
	end)
end
```

but you think this should be cached?

```
local function isBuilder(udef)
	return (udef.canReclaim and udef.reclaimSpeed > 0) or                         -- reclaim
		(udef.canResurrect and udef.resurrectSpeed > 0) or                        -- resurrect
		(udef.canRepair and udef.repairSpeed > 0) or                              -- repair
		(udef.buildOptions and udef.buildOptions[1]) or                           -- build options
		(udef.canStockpile and udef.modCategories.ship and udef.modCategories.noweapon) -- is carrier ship
end
```

WatchTheFort replied:
Yes

whossname replied:
I've added the `handleCaching` function to address this.

sprunk commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+				end
+
+				for _name, weapondef in pairs(udef.wDefs) do
+					if (weapondef.damages[vtoldamagetag] > weapondef.damages[defaultdamagetag]) then
```
damages for all categories are always defined

sprunk commented on luaui/Widgets/unit_key_select.lua:
Separating config and the widget proper lets people change the former without overriding the latter (which would make them miss out on functional improvements)

WatchTheFort replied:
Why is it disabled?

sprunk replied:
To let players do a wider live test before becoming the default

whossname replied:
It separates the functionality from the interface, allowing other widgets to use the functionality. I mostly did it to share common code between smart_select and key_select, but there should be other applications as well.

whossname commented:
I'll have a look this weekend. Could someone please clarify the workflow here? I haven't used PRs much. Am I meant to make the change and then mark the comment as resolved, or just reply to the comment?

whossname commented on luaui/Widgets/Include/select_api.lua:
I thought the first was what this did (based on how it behaved during debugging), but I can see the second is a better option. What is the error logging function for BAR?

WatchTheFort replied:
Search for use of `Spring.Log` and `LOG.ERROR`

whossname commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+			end
+
+			filters.name = invertCurry(invert, function(udef, _, _, name)
+				return stringContains(udef.translatedHumanName, name)
```
I had it that way originally, and agree that it isn't great. This is for backwards compatibility, in case there are people using this filter in their config.

WatchTheFort replied:
This is a buggy, fragile filter to even have, and we should remove it

whossname commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+
+local function parseConclusion(conclusionDef, commandDef)
+	local appendSelected = true
+	local prefix = conclusionDef:sub(1, 15)
```
This is just checking for the "ClearSelection_" prefix. SoC, this prefix is part of the conclusion, so it is handled here.

whossname commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+			uids = getCountUnits(uids, countUntil, appendSelected)
+			Spring.SelectUnitArray(uids, appendSelected)
+		end)
+	elseif startsWith(conclusionDef, "SelectPart_50+") then
```
yep, missed a pretty big one here.

whossname commented on luaui/Widgets/Include/select_api.lua:
issue is it's all pascal case, so setting it all to lower or upper case makes it harder to read

WatchTheFort replied:
It would still be pascal case in the code, but would accept any casing without error. This prevents bug reports when people use the incorrect case in their uikeys.txt file. Other languages have case insensitive string compare, but this is Lua, so...
```lua
local function tokenMatches(token, key)
	return lower(token) == lower(key)
end

if tokenMatches(token, "InPrevSel") or tokenMatches(token, "InPreviousSelection") then
```

whossname replied:
I've made filters, conclusions and sources case insensitive

whossname commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -0,0 +1,625 @@
+			end)
+		elseif token == "Patrolling" then
+			filters.patrolling = invertCurry(invert, function(_udef, _udefid, uid)
+				for i = 1, 4, 1 do
```
I don't know the reasoning, I just wanted to match the C++ code. I got the impression patrol might be implemented with a queue of sub commands. I looked into it a little bit at the time, but I wrote this code 3 months ago. I remember when I was debugging there always being a few commands in queue when a unit was patrolling, but I was looking at a few different things at the time, so not really sure.

WatchTheFort replied:
Add a comment saying why, but also see if this is actually needed, nothing worse than cargo cult programming.
```suggestion
				-- First four positions in queue checked because that is what the engine does
				for i = 1, 4, 1 do
```

whossname replied:
added the comment and did some testing. The patrol command is normally the second command in the queue, not the first command. This is when it is just patrolling, no other inputs. I don't know why it needs to check the first third and fourth positions in the queue, but it's safe to say that the patrol command is implemented using the command queue, which is the reason for this logic.

WatchTheFort commented on luaui/Widgets/Include/select_api.lua:
```diff
@@ -428,24 +411,12 @@ local function getCountUnits(uids, countUntil, appendSelected)
 
 	if not distance then
-		error("Invalid input: expected a number after the underscore.")
+		error("Invalid input, expected a number after the underscore: " .. input)
```
If `input` is nil, this itself will throw an error

WatchTheFort commented:
So apparently in a review, when you reply to a comment chain at the file level (i.e. not on a line of code), Github places the reply in the chain, but also adds a duplicate comment not connected to anything.

WatchTheFort commented:
@whossname Any status updates?

whossname commented:
Sorry, I've been busy. I probably won't get to this in the next month.

whossname commented:
I've finally got time to work on this. Addressed some of the comments and started testing. I'm getting a weird thing where I'm seeing warnings from the engine when the Widget should be overriding it. Has something changed? I don't remember it doing that before. The Widget is definitely working as expected otherwise, I'm just getting warnings from this line:

https://github.com/beyond-all-reason/spring/blob/782985a8e518d33b5af0636dbb078585967e214a/rts/Game/UI/SelectionKeyHandler.cpp#L393

whossname commented:
Just realised all my commits today were made from my work Git profile. damn.

WatchTheFort commented:
There's a better way to organize this using tables instead of long if-elseif, but let's see how this works.

WatchTheFort commented:
@whossname Can this be tested risk-free after merging, or does it need comprehensive testing before merge?

rhys-vdw commented:
> Just realised all my commits today were made from my work Git profile. damn.

You can fix it: https://stackoverflow.com/a/25815116/317135

whossname commented:
It should be risk free because the important stuff is disabled by default

On Thu, 16 Jan 2025, 4:39 am WatchTheFort, ***@***.***> wrote:

> @whossname <https://github.com/whossname> Can this be tested risk-free
> after merging, or does it need comprehensive testing before merge?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3757#issuecomment-2593888232>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXWLMM343VZ3ANOBRHPK3T2K3BQRAVCNFSM6AAAAABOUJB24GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJTHA4DQMRTGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WatchTheFort commented:
@whossname If you would like time to amend the commit history to change the author details, per rhys-vdw's comment, please let me know and I will hold off on merging until then. Otherwise, I am happy to merge this now.

whossname commented:
I'm happy for it to be merged as is. That mistake is annoying, but not a
huge issue for me.

Thanks for asking.

On Sat, 18 Jan 2025, 10:50 am WatchTheFort, ***@***.***>
wrote:

> @whossname <https://github.com/whossname> If you would like time to amend
> the commit history to change the author details, per rhys-vdw's comment,
> please let me know and I will hold off on merging until then. Otherwise, I
> am happy to merge this now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3757#issuecomment-2599486190>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXWLMOAKXQAICP653JC5FT2LG6QVAVCNFSM6AAAAABOUJB24GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGQ4DMMJZGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
